### ISSUE TITLE
Unexpected wake up when using suspend-then-hibernate

### ISSUE THREAD

**Issue by AlvinZhu** [systemd/systemd#24279]
### systemd version the issue has been seen with

251.4-1

### Used distribution

Arch Linux

### Linux kernel version used

5.18.16-arch1-1

### CPU architectures issue was seen on

x86_64

### Component

systemctl, systemd

### Expected behaviour you didn't see

After systemctl suspend-then-hibernate, system suspend(s2idle), after HibernateDelaySec=180min, system hibernate.

### Unexpected behaviour you saw

After systemctl suspend-then-hibernate, system automatically wakes up after 5minutes
Problem doesn't occur on regular suspend but only on suspend-then-hibernate

others has same issues:
https://bbs.archlinux.org/viewtopic.php?id=275965
https://unix.stackexchange.com/questions/705416/the-computer-wakes-up-every-5-minutes

my laptop is HP 845 G9 
CPU: AMD Ryzen 6850HS
S3 Sleep State is not support. Only Modern Standby(s2idle) and Hibernate works.

### Steps to reproduce the problem

run `systemctl suspend-then-hibernate`
then wait for 5 minutes.  system automatically wakes up

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
$ journalctl --no-pager -b 0 | grep -E "(systemd|kernel)"
Aug 11 18:54:52 Alvin-Laptop systemd[1]: Starting Network Manager Script Dispatcher Service...
Aug 11 18:54:52 Alvin-Laptop systemd[1]: Started Network Manager Script Dispatcher Service.
Aug 11 18:54:52 Alvin-Laptop kernel: wlp1s0: deauthenticating from ... by local choice (Reason: 3=DEAUTH_LEAVING)
Aug 11 18:54:52 Alvin-Laptop systemd-resolved[1039]: wlp1s0: Bus client reset search domain list.
Aug 11 18:54:52 Alvin-Laptop systemd-resolved[1039]: wlp1s0: Bus client set default route setting: no
Aug 11 18:54:52 Alvin-Laptop systemd-resolved[1039]: wlp1s0: Bus client reset DNS server list.
Aug 11 18:54:52 Alvin-Laptop systemd[1]: Reached target Sleep.
Aug 11 18:54:52 Alvin-Laptop systemd[1]: Starting RyzenAdj Service...
Aug 11 18:54:52 Alvin-Laptop systemd[1]: Starting Suspend; Hibernate if not used for a period of time...
Aug 11 18:54:52 Alvin-Laptop systemd[1]: ryzenadj.service: Deactivated successfully.
Aug 11 18:54:52 Alvin-Laptop systemd[1]: Finished RyzenAdj Service.
Aug 11 18:54:53 Alvin-Laptop systemd-sleep[9569]: bluetooth = off (software)
Aug 11 18:54:53 Alvin-Laptop systemd-sleep[9563]: Entering sleep state 'suspend'...
Aug 11 18:54:53 Alvin-Laptop kernel: PM: suspend entry (s2idle)
Aug 11 18:54:53 Alvin-Laptop kernel: Filesystems sync: 0.028 seconds
Aug 11 18:59:54 Alvin-Laptop kernel: Freezing user space processes ... (elapsed 0.002 seconds) done.
Aug 11 18:59:54 Alvin-Laptop kernel: OOM killer disabled.
Aug 11 18:59:54 Alvin-Laptop kernel: Freezing remaining freezable tasks ... (elapsed 0.001 seconds) done.
Aug 11 18:59:54 Alvin-Laptop kernel: printk: Suspending console(s) (use no_console_suspend to debug)
Aug 11 18:59:54 Alvin-Laptop kernel: ACPI: EC: interrupt blocked
Aug 11 18:59:54 Alvin-Laptop kernel: ACPI: EC: interrupt unblocked
Aug 11 18:59:54 Alvin-Laptop kernel: hpet_rtc_timer_reinit: 1039 callbacks suppressed
Aug 11 18:59:54 Alvin-Laptop kernel: hpet: Lost 9448 RTC interrupts
Aug 11 18:59:54 Alvin-Laptop kernel: [drm] PCIE GART of 512M enabled (table at 0x000000F4008CA000).
Aug 11 18:59:54 Alvin-Laptop kernel: amdgpu 0000:63:00.0: amdgpu: SMU is resuming...
Aug 11 18:59:54 Alvin-Laptop kernel: amdgpu 0000:63:00.0: amdgpu: dpm has been enabled
Aug 11 18:59:54 Alvin-Laptop kernel: amdgpu 0000:63:00.0: amdgpu: SMU is resumed successfully!
Aug 11 18:59:54 Alvin-Laptop kernel: i2c_hid_acpi i2c-ELAN0767:00: i2c_hid_get_input: incomplete report (14/65280)
Aug 11 18:59:54 Alvin-Laptop kernel: nvme nvme0: Shutdown timeout set to 10 seconds
Aug 11 18:59:54 Alvin-Laptop kernel: [drm] Watermarks table not configured properly by SMU
Aug 11 18:59:54 Alvin-Laptop kernel: nvme nvme0: 16/0/0 default/read/poll queues
Aug 11 18:59:54 Alvin-Laptop kernel: [drm] VCN decode and encode initialized successfully(under DPG Mode).
Aug 11 18:59:54 Alvin-Laptop kernel: [drm] JPEG decode initialized successfully.
Aug 11 18:59:54 Alvin-Laptop kernel: amdgpu 0000:63:00.0: amdgpu: ring gfx_0.0.0 uses VM inv eng 0 on hub 0
Aug 11 18:59:54 Alvin-Laptop kernel: amdgpu 0000:63:00.0: amdgpu: ring comp_1.0.0 uses VM inv eng 1 on hub 0
Aug 11 18:59:54 Alvin-Laptop kernel: amdgpu 0000:63:00.0: amdgpu: ring comp_1.1.0 uses VM inv eng 4 on hub 0
Aug 11 18:59:54 Alvin-Laptop kernel: amdgpu 0000:63:00.0: amdgpu: ring comp_1.2.0 uses VM inv eng 5 on hub 0
Aug 11 18:59:54 Alvin-Laptop kernel: amdgpu 0000:63:00.0: amdgpu: ring comp_1.3.0 uses VM inv eng 6 on hub 0
Aug 11 18:59:54 Alvin-Laptop kernel: amdgpu 0000:63:00.0: amdgpu: ring comp_1.0.1 uses VM inv eng 7 on hub 0
Aug 11 18:59:54 Alvin-Laptop kernel: amdgpu 0000:63:00.0: amdgpu: ring comp_1.1.1 uses VM inv eng 8 on hub 0
Aug 11 18:59:54 Alvin-Laptop kernel: amdgpu 0000:63:00.0: amdgpu: ring comp_1.2.1 uses VM inv eng 9 on hub 0
Aug 11 18:59:54 Alvin-Laptop kernel: amdgpu 0000:63:00.0: amdgpu: ring comp_1.3.1 uses VM inv eng 10 on hub 0
Aug 11 18:59:54 Alvin-Laptop kernel: amdgpu 0000:63:00.0: amdgpu: ring kiq_2.1.0 uses VM inv eng 11 on hub 0
Aug 11 18:59:54 Alvin-Laptop kernel: amdgpu 0000:63:00.0: amdgpu: ring sdma0 uses VM inv eng 12 on hub 0
Aug 11 18:59:54 Alvin-Laptop kernel: amdgpu 0000:63:00.0: amdgpu: ring vcn_dec_0 uses VM inv eng 0 on hub 1
Aug 11 18:59:54 Alvin-Laptop kernel: amdgpu 0000:63:00.0: amdgpu: ring vcn_enc_0.0 uses VM inv eng 1 on hub 1
Aug 11 18:59:54 Alvin-Laptop kernel: amdgpu 0000:63:00.0: amdgpu: ring vcn_enc_0.1 uses VM inv eng 4 on hub 1
Aug 11 18:59:54 Alvin-Laptop kernel: amdgpu 0000:63:00.0: amdgpu: ring jpeg_dec uses VM inv eng 5 on hub 1
Aug 11 18:59:54 Alvin-Laptop kernel: OOM killer enabled.
Aug 11 18:59:54 Alvin-Laptop kernel: Restarting tasks ... done.
Aug 11 18:59:54 Alvin-Laptop systemd-resolved[1039]: Clock change detected. Flushing caches.
Aug 11 18:59:54 Alvin-Laptop systemd-resolved[1039]: Switching to fallback DNS server 1.1.1.1#cloudflare-dns.com.
Aug 11 18:59:54 Alvin-Laptop systemd-sleep[9563]: System returned from sleep state.
Aug 11 18:59:54 Alvin-Laptop kernel: PM: suspend exit
Aug 11 18:59:54 Alvin-Laptop systemd[1]: Starting autorandr execution hook...
Aug 11 18:59:54 Alvin-Laptop systemd[1]: autorandr.service: Deactivated successfully.
Aug 11 18:59:54 Alvin-Laptop systemd[1]: Finished autorandr execution hook.
Aug 11 18:59:57 Alvin-Laptop systemd-sleep[9898]: bluetooth = on
Aug 11 18:59:57 Alvin-Laptop systemd[1]: systemd-suspend-then-hibernate.service: Deactivated successfully.
Aug 11 18:59:57 Alvin-Laptop systemd[1]: Finished Suspend; Hibernate if not used for a period of time.
Aug 11 18:59:57 Alvin-Laptop systemd[1]: Stopped target Sleep.
Aug 11 18:59:57 Alvin-Laptop systemd[1]: Reached target Suspend; Hibernate if not used for a period of time.
Aug 11 18:59:57 Alvin-Laptop systemd[1]: Stopped target Suspend; Hibernate if not used for a period of time.
Aug 11 18:59:57 Alvin-Laptop systemd-logind[1089]: Operation 'sleep' finished.
Aug 11 19:00:02 Alvin-Laptop systemd[1]: NetworkManager-dispatcher.service: Deactivated successfully.

$ cat /sys/power/mem_sleep   
[s2idle]
```

**Comment by AlvinZhu** [comment]
I fix this issue by replace the Qualcomm FastConnect 6900 WiFi card with Intel AX210.
and I also disabled WWAN(Fibocom L860-GL Intel XMM 7560) in BIOS, since it break suspend.

**Comment by renzhamin** [comment]
Yet another mediatek driver issue, suspend-then-hibernate works properly if we unload the driver beforehand hence a quick workaround with a systemd service file

```
echo \
"[Unit]
Before=suspend-then-hibernate.target
StopWhenUnneeded=yes

[Service]
Type=oneshot
StandardOutput=syslog
RemainAfterExit=yes
ExecStart=/bin/modprobe -r mt7921e
ExecStop=/bin/modprobe mt7921e

[Install]
WantedBy=suspend-then-hibernate.target" | \
sudo tee /etc/systemd/system/hack-suspend-then-hibernate.service &&
sudo systemctl enable hack-suspend-then-hibernate.service
```

**Comment by superm1** [comment]
Sorry to necro bump this bug, but I might know what's going on and I'd like if you guys can confirm it.

Those of you guys affected by this issue can you please experiment with `rtc_cmos.use_acpi_alarm=1` on the kernel command line?  Does it fix the issue?

**Comment by kishorv06** [comment]
@superm1 Thank you so much. You are a life saver. That works perfectly fine. I recently upgraded my laptop and I was so annoyed that laptop manufacturers got rid of s3 standby and the s2idle was a real pain for me. On top of that `suspend-then-hibernate` was also waking up every 5 mins draining the battery significantly. With this kernel parameter, it's fixed.

Could you please tell us what it does and how it fixed the issue ?

**Comment by tomasinoalvaro48** [comment]
Hardware: ASUS Zenbook 14 UM3402, AMD Ryzen 5 7530U with Radeon Graphics
OS: EndeavourOS (Arch Linux), kernel 7.0.10-arch1-1, systemd 260
Sleep state: [s2idle] (only available, no S3)
Bootloader: systemd-boot
Initramfs: dracut 109

Adding my device to this issue. suspend-then-hibernate does not wake automatically after HibernateDelaySec elapses — the system stays suspended indefinitely. When the lid is opened manually after the delay has passed, the system immediately hibernates instead of resuming normally. This confirms the RTC alarm is being set but never fires.

**/proc/driver/rtc:**
```
alarm_IRQ       : no
alrm_pending    : no
```

**/sys/power/mem_sleep:** [s2idle]

Tried kernel parameter `rtc_cmos.use_acpi_alarm=1` — no change. BIOS has no RTC wake alarm option.

Relevant journal excerpt (HibernateDelaySec=2min, lid closed at 21:04:09, opened manually at 21:06:28):
jun 21 21:04:09 systemd-logind: The system will suspend and later hibernate now!
jun 21 21:04:10 systemd-sleep: Performing sleep operation 'suspend'...
jun 21 21:04:10 kernel: PM: suspend entry (s2idle)
jun 21 21:04:11 kernel: PM: Some devices failed to suspend, or early wake event detected
jun 21 21:04:11 kernel: PM: suspend exit
jun 21 21:04:11 kernel: PM: suspend entry (s2idle)                                   ← system stays here indefinitely, no automatic wake after 2 min
jun 21 21:06:28 systemd-logind: Lid opened.                                              ← manual intervention
jun 21 21:06:28 systemd-sleep: System returned from sleep operation 'suspend-then-hibernate'.
jun 21 21:06:28 systemd-sleep: Performing sleep operation 'hibernate'...
jun 21 21:06:28 kernel: PM: hibernation: hibernation entry 

Hibernation itself works correctly — `systemctl hibernate` suspends and resumes perfectly.

Windows worked on this same hardware (hibernate after idle in sleep), so the capability exists but isn't exposed to Linux via the standard RTC alarm interface.

I also reported this issue on the EndeavourOS Forum: https://forum.endeavouros.com/t/suspend-then-hibernate-not-working-rtc-alarm-never-fires-on-asus-zenbook-14-um3402-ryzen-5-7530u-s2idle/80567

**Comment by superm1** [comment]
For 7530U, please update your kernel to pick up https://github.com/torvalds/linux/commit/f7ecfc3fe5753d10e9c4d0d7955c160ea8cbf0ea, hopefully it should help this issue.

**Comment by tomasinoalvaro48** [comment]
Thank you for the suggestion @superm1 . Unfortunately the issue persists on kernel 7.1.3-arch1-1.

I believe that commit may not apply to my hardware. My CPU is an AMD Ryzen 5 7530U (Barcelo-R, Zen 3+), not Cezanne — and the commit references amd_pmc_verify_czn_rtc which appears specific to Cezanne (czn).

Additionally, the amd_pmc driver does not appear to be active on my system at all.

alarm_IRQ is still no in /proc/driver/rtc after the kernel update, and the behavior is unchanged — the system stays suspended indefinitely after HibernateDelaySec elapses and only hibernates when the lid is opened manually.

Is there a different fix or workaround that would apply to Barcelo-R / Ryzen 7000 series?

**Comment by superm1** [comment]
Let's open a new issue and dig through specifics for your system.

Tag me in a new issue.
